compiler.rst

```diff
@@ -159,49 +159,81 @@ are very rare.  However, if you've allocated a PyObject, you must tell
-directly provided by the parse tree as in yacc.
+The function begins applying production rules on the tokens it
+receives; matching tokens recursively to their corresponding rule and calling
+the corresponding rule functions which usually follow the format :samp:`xx_rule`.
+Where *xx* is the grammar rule that the function handles (these names are
+derived from :file:`Grammar/python.gram`).
 
-For instance, one must keep track of which node in the parse tree
-one is working with (e.g., if you are working with an 'if' statement
-you need to watch out for the ':' token to find the end of the conditional).
+Each rule function in turn creates an AST node as it goes along.  It does this
+by allocating all the new nodes it needs, calling the proper AST node creation
+functions for any required supporting functions and connecting them as needed.
+This continues until all nonterminal symbols are replaced with terminals.  If an
+error occurs, the rule functions backtrack and try another rule function.  If
+there are no more rules, an error is set and the parsing ends.
 
-The functions called to generate AST nodes from the parse tree all have
-the name :samp:`ast_for_{xx}` where *xx* is the grammar rule that the function
-handles (``alias_for_import_name`` is the exception to this).  These in turn
-call the constructor functions as defined by the ASDL grammar and
-contained in :file:`Python/Python-ast.c` (which was generated by
-:file:`Parser/asdl_c.py`) to create the nodes of the AST.  This all leads to a
+The AST node creation functions have the name :samp:`_PyAST_{xx}` where *xx* is
+the AST node that the function creates. These are defined by the ASDL grammar
+and contained in :file:`Python/Python-ast.c` (which is generated by
+:file:`Parser/asdl_c.py` from :file:`Parser/Python.asdl`).  This all leads to a
 sequence of AST nodes stored in ``asdl_seq`` structs.
 
-Function and macros for creating and using ``asdl_seq *`` types as found
-in :file:`Python/asdl.c` and :file:`Include/asdl.h` are as follows:
+To avoid bad performance during backtracking, rules (other than the top level
+ones) are memoized.  Each rule checks if a memoized version exists and returns
+that if so, else it continues in the manner stated in the previous paragraph.
+
+Macros for using ``asdl_seq *`` types as found
+in :file:`Python/asdl.c` and :file:`Include/Internal/pycore_asdl.h` are as
+follows:
 
-``_Py_asdl_seq_new(Py_ssize_t, PyArena *)``
-        Allocate memory for an ``asdl_seq`` for the specified length
-``asdl_seq_GET(asdl_seq *, int)``
+``asdl_seq_GET_UNTYPED(asdl_seq *, int)``
         Get item held at a specific position in an ``asdl_seq``
-``asdl_seq_SET(asdl_seq *, int, stmt_ty)``
+``asdl_seq_SET_UNTYPED(asdl_seq *, int, stmt_ty)``
         Set a specific index in an ``asdl_seq`` to the specified value
 ``asdl_seq_LEN(asdl_seq *)``
         Return the length of an ``asdl_seq``
 
+There are also typed counterparts for some of the macros above.  These
```

Comment:
Let's mention that these sequence functions and another part of the API is automatically generated by the AST generation code, while the generic headers live in https://github.com/python/cpython/blob/main/Include/internal/pycore_asdl.h